### DISCUSSION TITLE
Help

### DISCUSSION THREAD

**Discussion by YuriiHauss** [codewars/codewars.com#3199] [General]
So below is my code in python for practice, can you tell me some advices to improve my code

import pickle

class University:
    def __init__(self, faculties_names, faculty_info, cathedra_info):
        self.faculties = faculties_names
        self.faculty_info = faculty_info
        self.cathedra_info = cathedra_info
        self.cathedra = None
        self.options = ["Викладачі", "Студенти"]
        self.option = None
        self.course1 = []
        self.course2 = []
        self.course3 = []
        self.course4 = []

    def choose_faculty(self):
        return choose_from_list(self.faculties)

    def choose_cathedra(self):
        faculty = self.choose_faculty()
        return None if not faculty else choose_from_list(self.faculty_info[faculty]["Кафедри"])

    def choose_option(self):
        places = ["Факультет", "Кафедру"]
        show_choices(places)
        choice = get_choice(len(places) + 1)
        if choice == len(places) + 1:
            return menu()
        options = [f"Створити інформацію про {places[choice - 1]}.", f"Видалити інформацію про {places[choice - 1]}.",
                   f"Редагувати інформацію про {places[choice - 1]}."]
        show_choices(options)
        option = get_choice(len(options) + 1)
        if choice == 1:
            faculty = self.choose_faculty()
            faculty = Faculty(faculty, self.faculties, self.faculty_info, self.cathedra_info)
            choices = [faculty.add_info, faculty.delete_info, faculty.edit_info, menu, menu]
            return choices[option - 1]()
        elif choice == 2:
            cathedra = self.choose_cathedra()
            if not cathedra:
                return
            cathedra = Cathedra(cathedra, self.faculties, self.faculty_info, self.cathedra_info)
            choices = [cathedra.add_info, cathedra.delete_info, cathedra.edit_info, menu, menu]
            return choices[option - 1]()

    def save_data(self):
        data = {
            "faculties": self.faculties,
            "faculty_info": self.faculty_info,
            "cathedra_info": self.cathedra_info,
        }
        with open("save_data.pickle", "wb") as f:
            pickle.dump(data, f)
        print("Інформацію збережено успішно")

    def read_data(self):
        try:
            with open("save_data.pickle", "rb") as f:
                data = pickle.load(f)
                self.faculties = data.get("faculties", [])
                self.faculty_info = data.get("faculty_info", {})
                self.cathedra_info = data.get("cathedra_info", {})
            print("Дані зчитано успішно, ви можете переглянути їх у текстовому файлі "
                  "після завершення роботи програми")
        except FileNotFoundError:
            print("Файл не знайдено, можливо, ви не зберігали нову інформацію")
        with open("saved_data.txt", "w", encoding="utf-8") as f:
            f.write("Faculties:\n")
            for faculty in self.faculties:
                f.write(f"- {faculty}\n")

            f.write("\nFaculty Info:\n")
            for faculty, info in self.faculty_info.items():
                f.write(f"- {faculty}:\n")
                for key, value in info.items():
                    f.write(f"  - {key}: {value}\n")

            f.write("\nCathedra Info:\n")
            for cathedra, info in self.cathedra_info.items():
                f.write(f"- {cathedra}:\n")
                for key, value in info.items():
                    f.write(f"  - {key}: {value}\n")

    def restore_data(self):
        with open("faculties.pickle", "rb") as f:
            self.faculties = pickle.load(f)
        with open("faculties_info.pickle", "rb") as f:
            self.faculty_info = pickle.load(f)
        with open("cathedras_info.pickle", "rb") as f:
            self.cathedra_info = pickle.load(f)
        print("Початкові дані відновлено успішно")

    def show_faculty_students_alph(self):
        all_students = []
        show_choices(self.faculties)
        index = get_choice(len(self.faculties) + 2)
        if 1 <= index <= len(self.faculties):
            self.faculty = self.faculties[index - 1]
        else:
            return menu()
        for i in self.cathedra_info:
            if i in self.faculty_info[self.faculty]["Кафедри"]:
                all_students.extend(self.cathedra_info[i]['Студенти'])
        all_students = sorted(all_students)
        print("Список студентів факультету за алфавітом: ")
        for student in all_students:
            print(student)

    def show_cathedra_stud_alph(self):
        all_students = []
        show_choices(self.faculties)
        index = get_choice(len(self.faculties) + 2)
        if 1 <= index <= len(self.faculties):
            self.faculty = self.faculties[index - 1]
        else:
            return menu()
        for i in self.cathedra_info:
            if i in self.faculty_info[self.faculty]["Кафедри"]:
                all_students.extend(self.cathedra_info[i]['Студенти'])
        all_students = sorted(all_students)
        print("Список студентів кафедри за алфавітом: ")
        for student in all_students:
            print(student)

    def show_faculty_teachers_alph(self):
        all_teachers = []
        show_choices(self.faculties)
        index = get_choice(len(self.faculties) + 2)
        if 1 <= index <= len(self.faculties):
            self.faculty = self.faculties[index - 1]
        else:
            return menu()
        for i in self.cathedra_info:
            if i in self.faculty_info[self.faculty]["Кафедри"]:
                all_teachers.extend(self.cathedra_info[i]['Викладачі'])
        all_teachers = sorted(all_teachers)
        print("Список викладачів кафедри за алфавітом: ")
        for teacher in all_teachers:
            print(teacher)

    def show_cathedra_teachers_alph(self):
        show_choices(self.faculties)
        index = get_choice(len(self.faculties) + 2)

        if 1 <= index <= len(self.faculties):
            self.faculty = self.faculties[index - 1]
        else:
            return menu()
        show_choices(self.faculty_info[self.faculty]["Кафедри"])
        index = get_choice(len(self.faculty_info[self.faculty]["Кафедри"]) + 2)
        if 1 <= index <= len(self.faculty_info[self.faculty]["Кафедри"]):
            self.cathedra = self.faculty_info[self.faculty]["Кафедри"][index - 1]
        else:
            return menu()
        teachers = self.cathedra_info[self.cathedra]["Викладачі"]
        teachers_sorted = sorted(teachers)
        print("Список викладачів кафедри за алфавітом: ")
        for teacher in teachers_sorted:
            print(teacher)

    def show_cathedra_stud_sorted_course(self):
        self.faculty = choose_from_list(self.faculties)
        if not self.faculty:
            return menu()
        cathedra = choose_from_list(self.faculty_info[self.faculty]["Кафедри"])
        if not cathedra:
            return menu()
        students_info = self.cathedra_info[cathedra]["Студенти"]
        students_by_course = {1: [], 2: [], 3: [], 4: []}
        for student, info in students_info.items():
            course = info[0]
            students_by_course[course].append(student)
        for course, students in sorted(students_by_course.items()):
            print(f"Курс {course}:")
            for student in sorted(students):
                print(student)

    def show_stud_alph_by_cathedra_and_course(self):
        self.faculty = choose_from_list(self.faculties)
        if not self.faculty:
            return menu()

        cathedra = choose_from_list(self.faculty_info[self.faculty]["Кафедри"])
        if not cathedra:
            return menu()

        courses = ["1 Курс", "2 Курс", "3 Курс", "4 Курс"]
        course_index = choose_from_list(courses)
        if not course_index:
            return menu()

        course = courses.index(course_index) + 1
        students_info = self.cathedra_info[cathedra]["Студенти"]
        students = [student for student, info in students_info.items() if info[0] == course]
        students_sorted = sorted(students)
        print("Список студентів кафедри за алфавітом і курсом: ")
        for student in students_sorted:
            print(student)

    def display_students_by_courses(self):
        course_students = {1: [], 2: [], 3: [], 4: []}

        for cathedra_info in self.cathedra_info.values():
            students_info = cathedra_info.get("Студенти", {})
            for student, student_info in students_info.items():
                course = student_info[0]
                course_students[course].append(student)

        for course, students in sorted(course_students.items()):
            print(f"{course} Курс:", sorted(students))

    def find_student(self):
        pib = input("Введіть ПІБ студента: ")
        for i in self.cathedra_info:
            for j in self.cathedra_info[i]["Студенти"]:
                if pib in j:
                    course = self.cathedra_info[i]["Студенти"][pib][0]
                    group = self.cathedra_info[i]["Студенти"][pib][1]
                    print(f"ПІБ: {pib}\nКафедра: {i}\nКурс: {course}\nГрупа: {group}")

    def find_teacher(self):
        pib = input("Введіть ПІб викладача: ")
        for i in self.cathedra_info:
            for j in self.cathedra_info[i]["Викладачі"]:
                if j == pib:
                    print(f"ПІБ:{j}\nКафедра:{i}")

    def show_stud_by_course_cathedra(self):
        options = ["Курс 1", "Курс 2", "Курс 3", "Курс 4"]
        show_choices(options)
        course_index = get_choice(len(options) + 2)

        if not 1 <= course_index <= len(options):
            return menu()
        course = course_index

        show_choices(self.faculties)
        faculty_index = get_choice(len(self.faculties) + 2)

        if not 1 <= faculty_index <= len(self.faculties):
            return menu()
        self.faculty = self.faculties[faculty_index - 1]

        show_choices(self.faculty_info[self.faculty]["Кафедри"])
        cathedra_index = get_choice(len(self.faculty_info[self.faculty]["Кафедри"]) + 2)

        if not 1 <= cathedra_index <= len(self.faculty_info[self.faculty]["Кафедри"]):
            return menu()
        self.cathedra = self.faculty_info[self.faculty]["Кафедри"][cathedra_index - 1]

        students_info = self.cathedra_info[self.cathedra].get("Студенти", {})
        students = [student for student, info in students_info.items() if info[0] == course]

        print(f"{self.cathedra}, {options[course - 1]}")
        for student in sorted(students):
            print(student)

class Faculty(University):
    def __init__(self, faculty, faculties_names, faculty_info, cathedra_info):
        super().__init__(faculties_names, faculty_info, cathedra_info)
        self.faculty = faculty
        self.info = None
        self.choice_info = None
        self.info_choices = ["Декан", "Кафедри", "Спеціальності"]

    def choose_faculty_info(self):
        try:
            self.info, self.choice_info = choose_info(self.faculty, self.faculty_info, self.info_choices)
        except TypeError:
            return

    def add_info(self):  # Якщо створюємо кафедру, вона створюється тільки в факулті інфо, треба ще щоб в кафеда інфо
        self.choose_faculty_info()
        if self.choice_info is None or self.choice_info == 4:
            return
        elif self.choice_info == 1:
            print("Неможливо додати інформацію про декана")
            return self.add_info()
        else:
            print(f"Існуючі {self.info_choices[self.choice_info - 1]}")
            show_choices(self.info)
        self.info.append(get_info(self.info_choices[self.choice_info - 1], self.info))
        print(self.faculty_info[self.faculty])

    def delete_info(self):
        self.choose_faculty_info()
        if self.choice_info == 4:
            return menu()
        elif self.choice_info == 1:
            print("Неможливо видалити інформацію про декана")
            return self.delete_info()
        else:
            print(f"Існуючі {self.info_choices[self.choice_info - 1]}")
            show_choices(self.info)
        choice = get_choice(len(self.info) + 2)
        if choice == len(self.info) + 1:
            return menu()
        if self.choice_info == 2:
            self.cathedra_info.pop(self.info[choice - 1])
        self.info.pop(choice - 1)
        print(self.faculty_info[self.faculty])

    def edit_info(self):
        self.choose_faculty_info()
        if self.choice_info == 1:
            info = get_info(self.info_choices[self.choice_info - 1], self.faculty_info[self.faculty]["Декан"])
            self.faculty_info[self.faculty].update({"Декан": info})
            print(self.faculty_info[self.faculty])
            return
        else:
            print(f"Існуючі {self.info_choices[self.choice_info - 1]}")
            show_choices(self.info)
        choice = get_choice(len(self.info) + 1)
        self.info[choice - 1] = get_info(self.info_choices[self.choice_info - 1], self.info[choice - 1])
        print(self.faculty_info[self.faculty])

class Cathedra(University):
    def __init__(self, cathedra, faculties_names, faculty_info, cathedra_info):
        super().__init__(faculties_names, faculty_info, cathedra_info)
        self.cathedra = cathedra
        self.info = None
        self.choice_info = None
        self.info_choices = ["Завідувач", "Спеціаліст кафедри", "Студенти", "Викладачі"]

    def choose_cathedra_info(self):
        try:
            self.info, self.choice_info = choose_info(self.cathedra, self.cathedra_info, self.info_choices)
        except TypeError:
            return

    def add_info(self):  # Хз як ще доставатися до курсів і груп
        self.choose_cathedra_info()
        if self.choice_info == 5 or self.choice_info is None:
            return
        elif self.choice_info in [1, 2]:
            print(f"Неможливо додати інформацію про {self.info_choices[self.choice_info - 1]}")
            return self.add_info()
        else:
            print(f"Існуючі {self.info_choices[self.choice_info - 1]}")
            show_choices(self.info)
        self.info.update({get_info(self.info_choices[self.choice_info - 1], self.info): []})
        print(self.cathedra_info[self.cathedra])
        return

    def delete_info(self):
        self.choose_cathedra_info()
        if self.choice_info == 5 or self.choice_info is None:
            return
        elif self.choice_info in [1, 2]:
            print(f"Неможливо видалити інформацію про {self.info_choices[self.choice_info - 1]}")
            return self.delete_info()
        else:
            print(f"Існуючі {self.info_choices[self.choice_info - 1]}")
            show_choices(self.info)
        choice = get_choice(len(self.info) + 1)
        if choice == len(self.info) + 1:
            return menu()
        self.info.pop(list(self.info)[choice - 1])
        print(self.cathedra_info[self.cathedra])

    def edit_info(self):
        self.choose_cathedra_info()
        if self.info_choices[self.choice_info - 1] == "Декан":
            info = input(
                f"Введіть ПІБ нового {self.info_choices[self.choice_info - 1]}а: ")
            self.cathedra_info[self.cathedra].update({"Декан": info})
            print(self.cathedra_info[self.cathedra])
            return
        else:
            print(f"Існуючі {self.info_choices[self.choice_info - 1]}")
            show_choices(self.choice)
        choice = get_choice(len(self.choice) + 2)
        info = input(f"Введіть інформацію про {self.info_choices[self.choice_info - 1]} на яку хочете змінити: ")
        self.choice[choice - 1] = info
        print(self.cathedra_info[self.cathedra])

def read(filename):
    with open(filename, 'rb') as f:
        return pickle.load(f)

university = University(read("faculties.pickle"), read("faculties_info.pickle"), read("cathedras_info.pickle"))

def menu():
    options = ["Створити/видалити/редагувати інформацію про факультет/кафедру [конкретного] факультета.",
               "Знайти інформацію про студента за ПІБ",
               "Знайти інформацію про викладача за ПІБ",
               "Вивести інформацію про всіх студентів впорядкованих за курсами.",
               "Вивести інформацію про всіх студентів факультета впорядкованих за алфавітом.",
               "Вивести інформацію про всіх викладачів факультета впорядкованих за алфавітом.",
               "Вивести інформацію про всіх студентів кафедри впорядкованих за курсами",
               "Вивести інформацію про всіх студентів кафедри впорядкованих за алфавітом.",
               "Вивести інформацію про всіх викладачів кафедри впорядкованих за алфавітом.",
               "Вивести інформацію про всіх студентів кафедри вказаного курсу.",
               "Вивести інформацію про всіх студентів кафедри вказаного курсу впорядкованих за алфавітом.",
               "Зберегти дані програми в локальне сховище.",
               "Відновити дані програми з локального сховища.",
               "Прочитати дані програми з локального сховища."]
    while True:
        print("Головне меню:")
        show_choices(options)
        choice = get_choice(len(options) + 2)
        choices = [university.choose_option, university.find_student,
                   university.find_teacher,
                   university.display_students_by_courses, university.show_faculty_students_alph,
                   university.show_faculty_teachers_alph,
                   university.show_cathedra_stud_sorted_course,
                   university.show_cathedra_stud_alph,  # Треба назви по менше, придумай щось
                   university.show_cathedra_teachers_alph,
                   university.show_stud_by_course_cathedra,
                   university.show_stud_alph_by_cathedra_and_course,
                   university.save_data,
                   university.restore_data,
                   university.read_data]
        if choice <= len(choices):
            return choices[choice - 1]()
        else:
            break

def get_choice(max_choice):
    choice = input(f"Оберіть опцію\n>>> ")
    try:
        if 1 <= int(choice) <= max_choice:
            return int(choice)
        else:
            print(
                f"\nНевірний вибір. Будь ласка, введіть число від 1 до {max_choice}.\n")  # Помилки б окремо було б з мега кайфом
            return get_choice(max_choice)
    except ValueError:
        print(f"\nНевірний вибір. Будь ласка, введіть число від 1 до {max_choice}.\n")
        return get_choice(max_choice)

def get_info(name, options):
    info = input(f"Введіть {name} які хочете додати: ")
    if info == str(len(options) + 1):
        return menu()
    elif info in options:
        print("Ця інформація вже збережена, введіть іншу")
        return get_info(name, options)
    return info

def show_choices(options):
    for index, option in enumerate(options):
        print(f"{index + 1}. {option}")
    print(f"{len(options) + 1}: Вийти в головне меню")

def choose_info(place, place_info, info_choices):
    show_choices(place_info[place])
    choice_info = get_choice(len(place_info[place]) + 1)
    if 1 <= choice_info <= len(info_choices):
        info = place_info[place][info_choices[choice_info - 1]]
        return info, choice_info
    return menu()

def main():
    while True:
        print("1. Показати головне меню")
        print("2.  Вийти")
        choice = get_choice(2)
        if choice == 2:
            break
        else:
            menu()

def choose_from_list(items):
    show_choices(items)
    index = get_choice(len(items) + 1)
    if 1 <= index <= len(items):
        return items[index - 1]
    return menu()

main()

**Comment by hobovsky** [comment]
Codewars Github is not meant for help with general programming tasks. It's meant to discuss matters related to Codewars platform, the website, its functions, etc.

For help with your assignment please visit your favorite homework forum, a code review forum, or you can try your luck asking for help on Codewars Discord.

Closing as off-topic.
